osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,237 @@
+
+  auto eids = context.constraints["eventid"].getAll(EQUALS);
+  if (!eids.empty()) {
+    xfilterList.emplace_back(
+        "(EventID=" + osquery::join(eids, ") or (EventID=") + ")");
+  }
+
+  auto pids = context.constraints["pid"].getAll(EQUALS);
+  if (!pids.empty()) {
+    xfilterList.emplace_back(
+        "(Execution[@ProcessID=" +
+        osquery::join(pids, "]) or (Execution[@ProcessID=") + "])");
+  }
+
+  auto times = context.constraints["time_range"].getAll(EQUALS);
+  auto timestamps = context.constraints["timestamp"].getAll(EQUALS);
+  if (!times.empty()) {
+    auto datetime = *times.begin();
+    auto time_vec = osquery::split(datetime, ";");
+
+    if (time_vec.size() == 1) {
+      auto _start = time_vec.front();
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start + "']");
+    } else if (time_vec.size() == 2) {
+      auto _start = time_vec.front();
+      auto _end = time_vec.at(1);
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start +
+                               "' and @SystemTime&lt;='" + _end + "']");
+    }
+  } else if (!timestamps.empty()) {
+    auto time_diff = *timestamps.begin();
+    xfilterList.emplace_back(
+        "TimeCreated[timediff(@SystemTime) &lt;= " + time_diff + "]");
+  }
+
+  xfilter = xfilterList.empty()
+                ? "*"
+                : "*[System[" + osquery::join(xfilterList, " and ") + "]]";
+}
+
+void genWindowsEventLog(RowYield& yield, QueryContext& context) {
+  std::set<std::pair<std::string, std::string>> xpath_set;
+
+  // Check if the `xpath` constraint is available. The `xpath` has
+  // priority over `channel` and it will lookup and filter the events
+  // based on `xpath` if passed.
+  if (context.hasConstraint("xpath", EQUALS)) {
+    auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+    auto xpath = *xpaths.begin();
+    pt::ptree propTree;
+    std::stringstream ss;
+    ss << xpath;
+    pt::read_xml(ss, propTree);
+    auto channel = propTree.get("QueryList.Query.Select.<xmlattr>.Path", "");
+    if (!channel.empty()) {
+      xpath_set.insert(std::make_pair(channel, xpath));
+    } else {
+      LOG(WARNING) << "Invalid xpath format - " << xpath;
+    }
+  } else if (context.hasConstraint("channel", EQUALS)) {
+    auto channels = context.constraints["channel"].getAll(EQUALS);
+    std::string xfilter("");
+    genXfilterFromConstraints(context, xfilter);
+    std::string welSearchQuery = kEventLogXmlPrefix;
+
+    for (const auto& channel : channels) {
+      welSearchQuery += "<Select Path=\"" + channel + "\">";
+      welSearchQuery += xfilter;
+      welSearchQuery += "</Select>" + kEventLogXmlSuffix;
+      xpath_set.insert(std::make_pair(channel, welSearchQuery));
+    }
+  } else {
+    LOG(WARNING) << "must specify the event log channel or xpath for lookup!";
+    return;
+  }
+
+  for (const auto& path : xpath_set) {
+    auto queryResults =
+        EvtQuery(nullptr,
```

Comment:
If I read this correctly, filtering can be done by either setting an xpath value, or by setting where clauses. As implemented, this looks like it takes each xpath value, and then iterates an `EvtQuery` for them, doing any subsequent filtering on the generated events. 

If that's correct, I do not think it is performant. 

The windows event log is huge. Fetching all the events to find things applicable to a single application is slow. And sqlite's query planner can make this especially bad.

Reply:
Ah, I see now. _If_ there is no xpath, `genXfilterFromConstraints` is called. This feels confusing.

Reply:
It's actually if there is no `xpath` and there is a `channel`. I added a comment above to see if we can still call `genXfilterFromConstraints` in the absence of both of these.

Reply:
The `channel` and `xpath` are required columns and the query should have at-least one of them. The `EvtQuery` function requires channel to be present and ignores it only if the channel/path information is present in the query string. 

In the absence of both, the query will not be successful and it will return with the error log.

Reply:
Error when neither is provided sounds good to me.